orttraining/orttraining/training_api/module.cc

```diff
@@ -43,13 +44,17 @@ std::unordered_set<const Node*> GetReverseReachableNodes(Graph& inference_graph,
 }
 
 Status RemoveUnusedNodes(Graph& inference_graph, InlinedVector<const NodeArg*>& output_node_args) {
-  auto reachable_nodes = GetReverseReachableNodes(inference_graph, output_node_args);
+  const auto reachable_nodes = GetReverseReachableNodes(inference_graph, output_node_args);
 
   // Get all graph nodes and remove those that are not in the reachable nodes.
   GraphViewer graph_viewer(inference_graph);
-  for (auto& node : graph_viewer.Nodes()) {
-    if (!reachable_nodes.count(&node)) {
-      inference_graph.RemoveNode(node.Index());
+  const auto node_indices = graph_viewer.GetNodesInTopologicalOrder();
+  for (size_t idx = node_indices.size(); idx > 0; --idx) {
+    const NodeIndex node_index = idx - 1;
```

Comment:
Would handling the size=0 case then iterate from size()-1 make it clearer?

Reply:
This was intentionally done to avoid a potential problem with `size_t`.

`for (size_t idx = 5; idx >= 0; idx--)` will run an infinite loop since `size_t` can never be < 0.

Reply:
After size = 0 is handled, size can only be >= 1 which is fine?

Reply:
Ah i see. Good catch. Although is there a conventionally cleaner way to handle it?

Reply:
ChatGPT suggested. Not sure if it’s good



#include <cstddef>
#include <iostream>

int main() {
    size_t count = 5;
    for (size_t i = count - 1; i != SIZE_MAX; --i) {
        std::cout << "Countdown: " << i << std::endl;
    }
    return 0;
}

Reply:
I had a similar solution in an [earlier commit](https://github.com/microsoft/onnxruntime/pull/15315/commits/e0c6a0797820de06403376b8c39e34ba502cc9ab):

```cc
for (NodeIndex node_index = node_indices.size()-1; node_index < node_indices.size(); --node_index)
```

This has 2 problems, 1). Not clear to the user what is going on, 2). MS SAL complains of ill defined for loop.

One other option that could be clearer is:

```cc
for (int64_t idx = static_cast<int64_t>(node_indices.size()), idx >= 0; --idx) {
    node_idx = static_cast<size_t>(idx);
    ...
}
```

But I think my current solution is cleaner in my opinion.